### PR TITLE
UCT/DEVX/SIG: Add signature attributes

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
## What
Add attributes describing signature offload buffer layout.

## How ?
```
worker_params.field_mask |= UCP_WORKER_PARAM_FIELD_FLAGS;
worker_params.flags      |= UCP_WORKER_FLAG_SIGNATURE;

worker_params.user_allocator.buffer_size = 4048;
worker_params.user_allocator.stride = 4096;
worker_params.user_allocator.offset = 0;
```

### CODE REVIEW

File: examples/ucp_client_server.c [hunk 1]
```diff
@@ -60,6 +59,7 @@ static sa_family_t ai_family   = AF_INET;
 static int num_iterations      = DEFAULT_NUM_ITERATIONS;
 static int user_allocator      = 0;
 static int sig                 = 0;
+static int sig_block           = 4048;
```
Comment (gleon99) [inline examples/ucp_client_server.c]: `const`? Maybe  even `#define`?

File: examples/ucp_client_server.c [hunk 2]
```diff
@@ -884,14 +879,21 @@ static ucs_mpool_ops_t memory_allocator_ops = {
...
-memory_allocator_init(ucp_context_h context, const size_t buffer_size,
+memory_allocator_init(ucp_context_h context,
                       ucp_worker_memory_allocator_obj_t **allocator_obj)
 {
     ucp_worker_memory_allocator_obj_t *allocator =
             (ucp_worker_memory_allocator_obj_t*)malloc(
                     sizeof(ucp_worker_memory_allocator_obj_t));
     ucs_mpool_params_t mp_params;
     ucs_status_t status;
+    size_t buffer_size;
```
Comment (gleon99) [inline examples/ucp_client_server.c]: Use ternary operator (simpler)?

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -1257,6 +1257,17 @@ typedef struct {
      * This will be the size of the active message fragment.
      */
     size_t                 buffer_size;
+
+    /**
+     * Stride in buffer provided by memory allocator. Should be great or
```
Comment (gleon99) [inline src/ucp/api/ucp.h]: `greater than, or equal to`

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -1085,6 +1085,28 @@ struct uct_iface_attr {
 };
 
 
+/*
+ * @ingroup UCT_RESOURCE
+ * @RX signature offload attributes
+ *
+ *  Attributes defining the way data provided by allocator will be
```
Comment (gleon99) [inline src/uct/api/uct.h]: Excessive indent?

File: src/uct/ib/mlx5/dv/sig.c [hunk 5]
```diff
@@ -266,17 +267,23 @@ ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
     size_t length;
     void *wqe;
 
+    if (sig_attr->block != 4048) {
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: Magic number

File: src/ucp/core/ucp_worker.c [hunk 6]
```diff
@@ -1343,8 +1334,24 @@ ucs_status_t ucp_worker_iface_open(ucp_worker_h worker, ucp_rsc_index_t tl_id,
     iface_params->field_mask |= UCT_IFACE_PARAM_FIELD_FEATURES;
     iface_params->features    = ucp_worker_get_uct_features(worker->context);
 
-    if (worker->flags & UCP_WORKER_FLAG_SIGNATURE) {
-        iface_params->features |= UCT_IFACE_FEATURE_SIGNATURE;
+    iface_params->field_mask |= UCT_IFACE_PARAM_FIELD_RX_HEADER_LENGTH;
+
+    if (worker->user_mem_allocator.cb != NULL) {
+        iface_params->field_mask |= UCT_IFACE_PARAM_FIELD_USER_ALLOCATOR |
+            UCT_IFACE_PARAM_FIELD_RX_PAYLOAD_LENGTH;
```
Comment (gleon99) [inline src/ucp/core/ucp_worker.c]: Indent?

File: src/ucp/core/ucp_worker.c [hunk 7]
```diff
@@ -1343,8 +1334,24 @@ ucs_status_t ucp_worker_iface_open(ucp_worker_h worker, ucp_rsc_index_t tl_id,
...
+        iface_params->rx_allocator.cb   = ucp_worker_user_allocator_get_cb;
+        iface_params->rx_payload_length =
+            worker->user_mem_allocator.buffer_size;
+
+        if (worker->flags & UCP_WORKER_FLAG_SIGNATURE) {
+            iface_params->features |= UCT_IFACE_FEATURE_SIGNATURE;
+            iface_params->rx_allocator.sig.block  =
```
Comment (gleon99) [inline src/ucp/core/ucp_worker.c]: Excessive spacing `  =`

File: src/uct/api/uct.h [hunk 8]
```diff
@@ -1099,6 +1121,9 @@ struct uct_rx_allocator {
 
     /* User-defined argument for the allocator callback */
     void                            *arg;
+
+    /* Signature offload attributes */
+    uct_sig_attr_t                  sig;
```
Comment (gleon99) [inline src/uct/api/uct.h]: `sig` -> `sig_attr`

File: src/uct/ib/mlx5/dv/sig.c [hunk 9]
```diff
@@ -257,7 +257,8 @@ static void* uct_ib_mlx5_sig_umr_start(uct_ib_mlx5_txwq_t *txwq,
 }
 
 ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
-                                     uct_ib_mlx5_mem_t *memh)
+                                     uct_ib_mlx5_mem_t *memh,
+                                     uct_sig_attr_t *sig_attr)
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: `const`

File: src/uct/ib/mlx5/dv/sig.c [hunk 10]
```diff
@@ -266,17 +267,23 @@ ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
...
+        return UCS_ERR_INVALID_PARAM;
+    }
+
     sig = ucs_malloc(sizeof(*sig), "sig");
     if (sig == NULL) {
         ucs_error("Cannot allocate signature context");
         return UCS_ERR_NO_MEMORY;
     }
 
+    sig->block = sig_attr->block;
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: Align `=`

File: src/uct/ib/mlx5/dv/sig.h [hunk 11]
```diff
@@ -9,15 +9,15 @@
...
-#define UCT_IB_MLX5_T10DIF_BLOCK  4048
-#define UCT_IB_MLX5_T10DIF_STRIDE 4096
-
 struct uct_ib_mlx5_sig {
     struct ibv_mr          *mr;
     struct mlx5dv_devx_obj *sig_mr;
     uint32_t               sig_key;
     void*                  *dif;
     struct ibv_mr          *dif_mr;
+    void*                  base;
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.h]: Maybe `start`? `start_addr`?

File: src/uct/ib/mlx5/dv/sig.h [hunk 12]
```diff
@@ -37,12 +38,17 @@ ucs_status_t uct_ib_mlx5_sig_mr_cleanup(uct_ib_mlx5_mem_t *memh);
 static inline unsigned uct_ib_mlx5_sig_mr_idx(uct_ib_mlx5_sig_t *sig,
                                               const void *ptr)
 {
-    return UCS_PTR_BYTE_DIFF(sig->mr->addr, ptr) / UCT_IB_MLX5_T10DIF_STRIDE;
+    unsigned idx = UCS_PTR_BYTE_DIFF(sig->base, ptr) / sig->stride;
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.h]: `const`

File: src/uct/ib/mlx5/dv/sig.h [hunk 13]
```diff
@@ -77,9 +83,9 @@ uct_ib_mlx5_calc_sig(uct_mem_h uct_memh, const void *data, const size_t len)
         uint16_t apptag;
         uint32_t reftag;
     } *t10dif               = UCS_PTR_BYTE_OFFSET(sig->dif, idx * 8);
-    size_t num_blocks       = len / UCT_IB_MLX5_T10DIF_BLOCK;
-    const uint16_t *p       = UCS_PTR_BYTE_OFFSET(data, UCT_IB_MLX5_T10DIF_BLOCK * num_blocks);
-    const uint16_t *end     = UCS_PTR_BYTE_OFFSET(p, (len % UCT_IB_MLX5_T10DIF_BLOCK) - 1);
+    size_t num_blocks       = len / sig->block;
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.h]: `const`?

File: examples/ucp_client_server.c [hunk 14]
```diff
@@ -887,12 +888,10 @@ memory_allocator_init(ucp_context_h context,
                     sizeof(ucp_worker_memory_allocator_obj_t));
     ucs_mpool_params_t mp_params;
     ucs_status_t status;
-    size_t buffer_size;
+    size_t buffer_size = ALLOCATOR_PAYLOAD_LENGTH;
```
Comment (gleon99) [inline examples/ucp_client_server.c]: 1. Unneeded init, in case `sig`. Maybe ternary?
2. Place initialized vars above uninitialized.

Comment (Artemy-Mellanox) [reply examples/ucp_client_server.c]: It's possible, but also it is good to keep optional features like user allocator, signature easy to localize, e.g. all signature related stuff under `if (sig) {}`

**Pull request merged**